components/builder-minio/habitat/config_install/minio-migration.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+
+function _hab_exec() {
+  pkg=$1
+  shift
+  hab pkg exec "$pkg" -- "$@"
+}
+
+function _aws() {
+  _hab_exec 'core/aws-cli' aws --endpoint-url "{{cfg.env.MINIO_ENDPOINT}}" "$@"
+}
+
+function _curl() {
+  _hab_exec 'core/curl' curl "$@"
+}
+
+function _jq() {
+  _hab_exec 'core/jq-static' jq "$@"
+}
+
+function download_bucket_objects() {
+  echo "DOWNLOADING objects from the MinIO that we are migrating from"
+  if ! _aws s3 sync "$s3_url" "$WAYPOINT"; then
+    echo "ERROR: download of objects FAILED"
+    exit 8
+  fi
+  return 0
+}
+
+function _ensure_bucket_exists() {
+  if ! _aws s3 ls "$MINIO_BUCKET" &>/dev/null; then
+    _aws s3 mb "$s3_url"
+  fi
+}
+
+function upload_bucket_objects() {
+  _ensure_bucket_exists
+  echo "UPLOADING objects to the MinIO that we are migrating to"
+  if ! _aws s3 sync "$WAYPOINT" "$s3_url"; then
+    echo "ERROR: upload of objects FAILED"
+    exit 9
+  fi
+  return 0
+}
+
+function is_migration_from_removed_fs_backend_needed() {
+  echo "CHECKING if migration from removed fs backend is needed"
+  if [[ -f /hab/svc/builder-minio/data/.minio.sys/format.json ]]; then
+    format_value=$(_jq -r '.format' /hab/svc/builder-minio/data/.minio.sys/format.json)
+    if [[ "${format_value}" == 'fs' ]]; then
+      return 0
+    fi
+  fi
+  return 1
+}
+
+function minio_health_live_check() {
+  for ((n = 0; n < 20; n++)); do
+    curl_http_code_args=(-fs -o /dev/null -w "%{http_code}" --retry 4 --retry-delay 1)
+    code=$(_curl "${curl_http_code_args[@]}" "$MINIO_ENDPOINT/minio/health/live")
+    if [[ $code == 200 ]]; then
+      return 0
+    else
+      sleep .5
+    fi
+  done
+  return 1
+}
+
+function minio_stop() {
+  for ((n = 0; n < 20; n++)); do
+    if pgrep minio >/dev/null; then
+      pkill minio >/dev/null
+      sleep 1
+    else
+      return 0
+    fi
+  done
+}
+
+function config_environment_for_migration() {
+  if [[ ! $1 =~ ^[0-9]+$ ]]; then
+    echo "ERROR: Invalid timestamp"
+  fi
+  export MINIO_ENDPOINT="{{cfg.env.MINIO_ENDPOINT}}"
```

Comment:
maybe use `https://localhost:{{cfg.bind_address}}` like the other hooks do.

Reply:
Since this will execute in the context of bldr.env and our users might have changed that there shouldn't we keep it?  I was going to change it and align with the other hooks but that was nagging at me. My counter to myself on that is if people had been changing it then why is hook-helper.sh broken?  I think the answer to that question and what answer your question posed in these three items boils down to the user.toml files created [here](https://github.com/habitat-sh/on-prem-builder/blob/d5701f5c5fbc8948cde2ef00e1207e0e3cf1d2fb/scripts/provision.sh#L70) and [here](https://github.com/habitat-sh/on-prem-builder/blob/d5701f5c5fbc8948cde2ef00e1207e0e3cf1d2fb/scripts/provision.sh#L98).  I think that explains the difference in the bucket name (next comment) and and why we should look to the environment.

Reply:
The `MINIO_ENDPOINT` in bldr.env is only used if they are using aws S3 or their own external s3. It's not used if they are using our packaged minio.

Reply:
Done.